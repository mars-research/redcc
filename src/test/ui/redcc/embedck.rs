// build-pass

#![feature(rustc_attrs)]

#[cfg_attr(not(test), rustc_diagnostic_item = "RRef")]
struct RRef<T>(T);
struct ContainsRRefDirect(RRef<i32>);

// Detect trivial embedding and non-embedding assignments
fn main() {
    let r = RRef(4);
    let mut rc = RRef(ContainsRRefDirect(RRef(1)));
    rc.0.0 = r;

    let r = RRef(4);
    let mut c = ContainsRRefDirect(RRef(1));
    c.0 = r;
}

// more test cases:
    // Indirect embedding
    // Sibling RRef does not trigger embed
    // embedding with different kinds of rvalues
        // enums too
    // embedding with different kinds of places
        // array
        // tuple
        // ref/pointer
        // enum
    // rvalue with embedded rref works
