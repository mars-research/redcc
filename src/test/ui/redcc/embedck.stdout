EMBED RRef<i32> -> RRef<ContainsRRefDirect>
